sdks/python/apache_beam/transforms/core.py

```diff
@@ -459,6 +474,50 @@ def reset(self):
     del self._callbacks[:]
 
 
+class WatermarkEstimator(object):
+  """A WatermarkEstimator which is used for tracking output_watermark per <>in a
+  DoFn.process(), typically tracking per <element, restriction> pair in SDF in
+  streaming.
+
+  There are 3 APIs in this class: update_watermark, current_watermark and reset
```

Comment:
I think we should flesh out what the interface is for a general WatermarkEstimator, vs. this particular ManualWatermarkEstimator with this particular implementation. This may inform whether we want a WatermarkEstimatorProvider or an actual WatermarkEstimator instance (which we repeatedly call reset() on.) A TODO would be fine.

Reply:
Created [BEAM-8537](https://issues.apache.org/jira/browse/BEAM-8537). I'll follow up with this one after merging this PR.